spec/todoApp.spec.js

```diff
@@ -0,0 +1,175 @@
+const todoApp = require("../src/todoApp");
+
+describe("Todo app", () => {
+  afterEach(function () {
+    todoApp.nextTodoId = 0;
+    todoApp.todoList = [];
+  });
+
+  it("creates a todoItem", () => {
+    // setup
+    const text = "learn TDD";
+    // execute
+    const result = todoApp.create(text);
+    //verify
+    expect(result).toEqual({
+      id: 1,
+      description: text,
+      status: "incomplete",
+    });
+  });
+
+  it("increments IDs by one when adding new item", () => {
+    // setup
+    const task1 = "write tests";
+    const task2 = "write tests";
+    // execute
+    const result1 = todoApp.create(task1);
+    const result2 = todoApp.create(task2);
+    //verify
+    expect(result1).toEqual({
+      id: 1,
+      description: task1,
+      status: "incomplete",
+    });
+    expect(result2).toEqual({
+      id: 2,
+      description: task2,
+      status: "incomplete",
+    });
+  });
+
+  it("returns error message if no description provided", () => {
+    // setup
+    const text = "";
+    // execute
+    const result = todoApp.create(text);
+    //verify
+    expect(result).toEqual("Error: you need to specify the description!!!");
+  });
+
+  it("shows all todo items if the array is not empty", () => {
+    //   it("method showAll() returns array with todos when todos array has at least one item in it", () => {
+    // setup
+    const task1 = "write tests";
+    const todo1 = todoApp.create(task1);
+    const todoList = [todo1];
+    // execute
+    const result = todoApp.showAll();
+    //verify
+    expect(result).toEqual(todoList);
+  });
+
+  it("shows error message if the array is empty", () => {
+    //it("method showAll() returns error message if the todos array is empty", () => {
+
+    // setup
+    const errorMessage = "There is nothing to do in this array...";
+    // execute
+    const result = todoApp.showAll();
+    //verify
+    expect(result).toEqual(errorMessage);
+  });
+
+  it("method setComplete() sets todo item's status to complete", () => {
+    // setup
+    const task1 = "write tests";
+    const todo1 = todoApp.create(task1);
+    // execute
+    const result = todoApp.setComplete(todo1.id);
+    //verify
+    expect(result.status).toEqual("complete");
+  });
+
```

Comment:
So do lines 84-96 for example switch task1 to todo1 when you run the setComplete(task1)?

Reply:
Here on line 86 I setup the text for description and store it in `task1` variable. Then on line 89 I'm passing it as an argument to `todoApp.create()` function call, which creates an object like so: `{id: 1, description: "write tests", status: "incomplete"}`. Also I'm setting up other 2 variables same way(that could be improved, because here I'm repeating same process 3 times). On line 92 I'm calling `todoApp.setComplete()` which takes an id as an argument and changes the status of `todo2` in this case from `incomplete` to `complete`. So after this I have `todo1` and `todo3` objects with `status: "incomplete"`, and `todo2` has `status: "complete"`. On line 93 I'm setting up my expected result what I should get after calling `todoApp.showIncomplete()`(which should be equal to my expected result, which I'm storing in `incompleteTodoList` variable).

Reply:
If I pass `todoApp.setComplete(task1)` here, my test will fail, because `task1` is a string, while `setComplete()` is expecting an `id`. So if I want to change the state of `todo1`(which is an object), I need to pass its' id to the function call, in this case - `todoApp.setComplete(todo1.id)`.